src/plugins/tlsmon/tlsmon.cpp

```diff
@@ -0,0 +1,293 @@
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+    // We can validate that we indeed found NCryptSslKey by checking magic
+    // bytes value.
+    if (ncrypt_ssl_key.magic != tlsmon_priv::NCRYPT_SSL_KEY_MAGIC_BYTES)
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // NCryptSslKey contains a pointer to SslMasterSecret structure.
+    tlsmon_priv::__ssl_master_secret_t master_secret;
+    ctx.addr = (addr_t) ncrypt_ssl_key.master_secret;
+    if (VMI_SUCCESS != vmi_read(lg.vmi, &ctx, sizeof(master_secret), &master_secret, nullptr))
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+    // Again we can validate that we found SslMasterSecret structure by
+    // checking magic bytes.
+    if (master_secret.magic != tlsmon_priv::MASTER_SECRET_MAGIC_BYTES)
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+
+    // Now retrieve client random value. pParameterList points to an array of
+    // NCryptBuffer buffers which contains at least client and server random
+    // values.
+    ctx.addr = params->parameter_list_addr;
+    tlsmon_priv::__ncrypt_buffer_desc_t ncrypt_buffer_desc;
+    if (VMI_SUCCESS != vmi_read(lg.vmi, &ctx, sizeof(ncrypt_buffer_desc), &ncrypt_buffer_desc, nullptr))
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    std::vector<tlsmon_priv::__ncrypt_buffer_t> ncrypt_buffers = std::vector<tlsmon_priv::__ncrypt_buffer_t>(ncrypt_buffer_desc.cbuffers);
+    ctx.addr = (addr_t) ncrypt_buffer_desc.buffers;
+    if (VMI_SUCCESS != vmi_read(lg.vmi, &ctx, ncrypt_buffer_desc.cbuffers * sizeof(tlsmon_priv::__ncrypt_buffer_t), ncrypt_buffers.data(), nullptr))
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // Find the buffer containing client random.
+    auto it = std::find_if(ncrypt_buffers.begin(), ncrypt_buffers.end(), [&](const auto& e)
+    {
+        return e.buffer_type == tlsmon_priv::NCRYPTBUFFER_SSL_CLIENT_RANDOM;
+    });
+    if (it == ncrypt_buffers.end())
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // And finally read it.
+    std::array<char, tlsmon_priv::CLIENT_RANDOM_SZ> client_random = std::array<char, tlsmon_priv::CLIENT_RANDOM_SZ>();
+    ctx.addr = (addr_t) it->buffer;
+    if (VMI_SUCCESS != vmi_read(lg.vmi, &ctx, client_random.size(), client_random.data(), nullptr))
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+
+    // Output retrieved data in hex format.
+    std::string master_key_str = tlsmon_priv::byte2str((unsigned char*)master_secret.master_key, tlsmon_priv::MASTER_KEY_SZ);
+    std::string client_random_str = tlsmon_priv::byte2str((unsigned char*)client_random.data(), tlsmon_priv::CLIENT_RANDOM_SZ);
```

Comment:
hmmm, if these 2 are being cast, shouldn't they just be `unsigned char*` in structs?

Reply:
I've changed `master_secret.master key` field type to `unsigned char*`. However, `client_random` is `std::string`.